frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/object-with-type-property-wrapper.ts

```diff
@@ -1,82 +1,86 @@
-          : initializeWithTypeData
-        : this.getTypeData(initializeWithType);
-      this.setType(initializeWithType, resultTypeData);
-    }
   }
 
-  getType = (): TYPE =>
+  public getType = (): TYPE =>
     this.TypeClass.getAll().find((type) => this.getIn([...this.typeDataPath, type.getValue()]));
 
-  getTypeValue = (): string => {
+  public getTypeValue = (): string => {
     const type = this.getType();
     return type && type.getValue();
   };
 
-  hasType = (): boolean => !!this.getType();
+  public hasType = (): boolean => !!this.getType();
 
-  protected getTypeData = (type?: TYPE) =>
+  public getTypeData = (type?: TYPE): COMBINED_TYPE_DATA =>
     this.getIn([...this.typeDataPath, (type || this.getType()).getValue()]);
 
-  protected setType = (type?: TYPE, typeData?: any) => {
+  public mergeWith(...wrappers: SELF[]): SELF {
+    super.mergeWith(...wrappers);
+    const lastWithType = _.last(wrappers.filter((wrapper) => wrapper?.getType()));
+
+    if (lastWithType) {
+      this.appendType(lastWithType.getType(), undefined, false); // removes typeData of other types
+    }
+    return (this as any) as SELF;
+  }
+
+  public setType = (type?: TYPE, typeData?: COMBINED_TYPE_DATA, sanitize = true) => {
     const typeDataParent =
       this.typeDataPath.length === 0 ? this.data : this.getIn(this.typeDataPath);
-    if (!typeDataParent) {
-      return;
-    }
-    this.TypeClass.getAll().forEach(
-      (superflousProperty) => delete typeDataParent[superflousProperty.getValue()],
-    );
-    if (type) {
-      typeDataParent[type.getValue()] = typeData ? _.cloneDeep(typeData) : {};
+    if (typeDataParent) {
+      this.TypeClass.getAll().forEach(
+        (superfluousProperty) => delete typeDataParent[superfluousProperty.getValue()],
+      );
+      if (type) {
+        const finalTypeData = typeData
+          ? sanitize
+            ? this.sanitize(type, typeData) || {}
+            : _.cloneDeep(typeData)
+          : {};
+        if (sanitize) {
+          omitEmpty(finalTypeData, true);
+        }
+        typeDataParent[type.getValue()] = finalTypeData;
+      }
     }
+    return (this as any) as SELF;
   };
 
-  protected addTypeData = (newTypeData?: any) => {
-    const type = this.getType();
-    this.setType(type, { ...this.getTypeData(type), ...newTypeData });
-  };
+  public appendType = (type?: TYPE, newTypeData?: COMBINED_TYPE_DATA, sanitize = true) =>
+    this.setType(type, { ...this.getTypeData(type), ...newTypeData }, sanitize);
+
+  public setTypeData = (newTypeData?: COMBINED_TYPE_DATA, sanitize = true) =>
+    this.setType(this.getType(), newTypeData, sanitize);
+
+  public appendTypeData(newTypeData?: COMBINED_TYPE_DATA, sanitize = true) {
+    return this.appendType(this.getType(), newTypeData, sanitize);
+  }
+
+  // should be implemented by derived wrappers
+  protected sanitize(type: TYPE, typeData: COMBINED_TYPE_DATA): any {
+    return _.cloneDeep(typeData);
```

Comment:
:+1: on having a sensible fallback implementation.

An alternative is to mark the method as `abstract` which will force sub-classes to implement it, but this adds more potential code to these sub-classes.

Reply:
shouldn't hurt having this here for now